core/src/main/scala/kafka/tools/StorageTool.scala

```diff
@@ -60,7 +63,11 @@ object StorageTool extends Logging {
           if (!metadataVersion.isKRaftSupported) {
             throw new TerseFailure(s"Must specify a valid KRaft metadata version of at least 3.0.")
           }
-          val metaProperties = buildMetadataProperties(clusterId, config.get)
```

Comment:
@cmccabe Sorry for making noise in this merged PR, but I have a question about valid "cluster id"

Before this patch, the `cluster id` was required to be a "valid" UUID (`buildMetadataProperties` does validation). For example, the length must be larger than 24. However, this patch removed the validation of UUID and hence we can start a cluster with non-uuid cluster id. Is this change expected?

Reply:
I'm keen to get @cmccabe to confirm, but suspect this is not intentional.

Since we released 3.7, 3.7.1 and soon 3.8 without this validation, we should continue to accept it in the broker startup logic so we don't break anyone's existing cluster. But I don't see a reason why we shouldn't add the validation back to StorageTool, to prevent new clusters being created with a non UUID value for cluster ID. WDYT?

Reply:
> Since we released 3.7, 3.7.1 and soon 3.8 without this validation, we should continue to accept it in the broker startup logic so we don't break anyone's existing cluster. But I don't see a reason why we shouldn't add the validation back to StorageTool, to prevent new clusters being created with a non UUID value for cluster ID. WDYT?

agree that adding uuid check back will have compatibility issue. Also, I grep code base and don't find any uuid restriction to cluster id. Hence, I plan to open a jira to address following changes:

1.  update docs (https://kafka.apache.org/documentation/#kraft_storage) to remind that "cluster id" can be non-uuid after 3.7
2. change the type from `uuid` to `string` for testing code to avoid misunderstand.
3. cleanup the unused code from `StroageTool` (https://issues.apache.org/jira/browse/KAFKA-17118)

Reply:
If this was an accident, I don't think we should make it permanent.

Reply:
> I don't think we should make it permanent.

If the "uuid check" is added back to `StorageTool`, we have to face following issues:

1. compatibility (as @soarez mentioned)
2. the conflicts caused by base64. That happens recently in our cluster.  the input cluster id is "lYEJ7pnYNO2ovNYiNRJBGl", but the cluster id stored in `meta.properties` is "lYEJ7pnYNO2ovNYiNRJBGg".

Reply:
> If this was an accident, I don't think we should make it permanent.

@ijuma could you clarify your suggestion a bit? Do you mean that we shouldn't make it permanent:

1. only in the generation of meta.properties (StorageTool)
2. also in the loading of meta.properties (KafkaRaftServer)
3. (some other option?)

If you're suggesting we re-introduce loading validation (2), are there any options for operators that have created clusters in production with cluster IDs that would fail that validation? Or should we not mind those?

Reply:
I feel the most important thing is "Had we ever convert cluster id from string to UUID"? If the answer is YES, we ought to add the check back to `StorageTool` and then announce that the cluster id written by 3.7.0 and 3.7.1 is not validated according to UUID format.

Fortunately, I don't notice that usage (string -> UUID) currently.  Hence, the compatibility issue happens in `StorageTool`: Should we allow user to use non-uuid cluster id to format? If no, we need to handle the compatibility issue and the conflicts of base64 encoding. If yes, the side effect is ... user will use shorter/longer string? Sorry for my poor imagination :cry: 

At any rate, we are going to release 3.8.0. Hence, it would be nice to have a decision ASAP.

Reply:
The KRaft formatting tool did have a requirement that the cluster ID had to be a Kafka UUID for a while. However, cluster ID has always been a string when in ZK mode. Because we now support migration from ZK to KRaft, we need to support cluster ID as a string.

If we want to have cluster IDs be UUIDs then I think it would be better to have the discussion on the mailing list than here in this PR. The main question we'd have to answer is how to convert existing string cluster IDs to UUID. Another option, I suppose, would be to require a UUID for new clusters, but allow the arbitrary string for old ones. The main advantage of that would probably just be preventing pathological cluster names (people trying to name their cluster a whitespace string, or an emoji, etc.)

I don't think we will be able to get to the point where the APIs for fetching cluster ID return UUID very easily... it will be a multi-year effort. Unfortunately it's very easy to create string IDs, and very hard to get rid of them.

Reply:
That makes sense - so this is not an accidental change as it was stated earlier.

Reply:
@cmccabe thanks for great explanation. We will leverage your comments to update the docs of StorageTool!